### PR TITLE
Reconnect Notice: In-place Reconnect

### PR DESCRIPTION
**Pull request by sergeymitr**
#### Changes proposed in this Pull Request:
This PR introduces a reconnection flow, which allows users to reconnect Jetpack in-place, in case of connection errors.
In particular, this PR:
- Adds a `connection/reconnect` endpoint, that accepts `POST` requests, with a **required** `action` parameter. The `action` parameter for now can be only `reconnect`. This corresponds to re-registering the site and deleting all tokens (user and blog), so the users need to re-authorize wpcom and re-issue their blog tokens. In the future, this endpoint will also support the actions: `refresh_blog_token` and `refresh_user_token`. 
- Displays an error notice, in cases of connection errors, along with a `Restore connection` action prompt, that triggers the reconnection flow
- Displays the authorization iframe so that a user can re-connect their wpcom account in-place, as part of the reconnection flow (after successful response from the `connection/reconnect` endpoint the user needs to re-connect because the corresponding user tokens are now deleted). By setting the `JETPACK_SHOULD_NOT_USE_CONNECTION_IFRAME` constant  `true` or using **Safari** browser, instead of displaying the authorization iframe the user is redirected to wpcom in order to authorize their account.
- Moves all related Admin page files to using Core translation functions in the context of [Admin Page: Move from i18n-calypso to Core functions to handle translations](https://github.com/Automattic/jetpack/issues/16481)

#### Jetpack product discussion
`p9dueE-1CF-p2`

#### Does this pull request change what data or activity we track or use?
No.

#### Testing instructions:

**Prerequisites:**
- A "broken" Jetpack connection: Using the Broken Token plugin, invalidate the **blog token**.

**Testing:**
- Make sure after "breaking" the blog token, the connection error is visible along with the `Restore connection` CTA.
- Click the `Restore connection` CTA: The admin UI should blur and you should see an info message that reconnection is in progress
- The authorization iframe should be displayed in order to re-connect your wpcom account
- After re-connecting your wpcom account the page will refresh and your connection should be now healthy (the connection error shouldn't display anymore)

**Note:** By setting the `JETPACK_SHOULD_NOT_USE_CONNECTION_IFRAME` constant  `true` or testing this with **Safari** browser, instead of seeing the authorization iframe you should be redirected to wpcom in order to authorize your account.

**Before:**
N/A

**After:**
https://d.pr/v/oN8gUa

#### Proposed changelog entry for your changes:
* Easy reconnect in case of connection errors.

### CODE REVIEW

**Comment by jeherve** [comment]
As you work on this, do you think you could start using `@wordpress/i18n` for all new translatable strings instead of the existing `i18n-calypso`? This should save us a bit of rebase troubles as we make progress on #16481.

File: _inc/client/state/connection/actions.js [hunk 1]
```diff
@@ -309,11 +315,11 @@ export const reconnectSite = ( action = 'reconnect' ) => {
...
 						'is-error',
-						__( 'There was an error reconnecting Jetpack. Error: %(error)s', {
-							args: {
-								error: error,
-							},
-						} ),
+						sprintf(
+							/* translators: placeholder is the error. */
+							__( 'There was an error reconnecting Jetpack. Error: %s', 'jetpack' ),
+							error.response.message || error.response.code
```
Comment (fgiannar) [inline _inc/client/state/connection/actions.js:321]: @sergeymitr @jeherve fyi this was accidentally [reverted in a previous commit ](https://github.com/Automattic/jetpack/pull/16475/commits/bbd297688a0ba4eb44e692ae71dd9bacfb0662d8)and took the opportunity to fix it here

Comment (sergeymitr) [reply _inc/client/state/connection/actions.js:321]: Good catch, thanks 👍

**Review by kbrown9 — CHANGES_REQUESTED** [review]
This works well for me. I tested with the iframe and Calypso flow. I just have a couple of comments about the code.

File: _inc/client/components/jetpack-notices/jetpack-connection-errors.jsx [hunk 2]
```diff
@@ -78,3 +45,5 @@ export default class JetpackConnectionErrors extends React.Component {
 		return Object.values( errorsToDisplay ).map( error => this.renderOne( error ) );
```
[the remark this answers is not in the document]
Comment (sergeymitr) [reply _inc/client/components/jetpack-notices/jetpack-connection-errors.jsx]: TIL that `React.Fragment` objects require `key` attributes when in an array.
Thanks Kim 🙂

File: packages/connection/src/class-rest-connector.php [hunk 3]
```diff
@@ -91,18 +91,35 @@ public function __construct( Manager $connection ) {
...
+			array(
+				'methods'  => WP_REST_Server::EDITABLE,
+				'callback' => array( $this, 'connection_reconnect' ),
+				'args'     => array(
+					'action' => array(
+						'type'     => 'string',
+						'required' => true,
+					),
+				),
+				'permission_callback' => '__return_true',
```
[the remark this answers is not in the document]
Comment (sergeymitr) [reply packages/connection/src/class-rest-connector.php]: Callback added, thanks for pointing that out.

File: _inc/client/components/jetpack-notices/jetpack-connection-errors.jsx [hunk 4]
```diff
@@ -78,3 +45,5 @@ export default class JetpackConnectionErrors extends React.Component {
...
 	}
 }
+
+export default connect( null, null )( JetpackConnectionErrors );
```
[the remark this answers is not in the document]
Comment (sergeymitr) [reply _inc/client/components/jetpack-notices/jetpack-connection-errors.jsx]: It's probably a leftover, good catch.
Removed.

**Review by kbrown9 — APPROVED** [review]
The changes look good to me!

I tested with both the iframe and Calypso reconnect flows, and everything worked well. I also sent an `action: reconnect` POST request to the reconnect endpoint, and the request returned with a 401 status and an invalid user permission message, as expected.

**Pull request merged**